File: haohaninfo/交易實務案例/order.R
#設定下單程式位置
ExecPath <- "./bin/"

#市價委託單
OrderMKT<-function(Product,BorS,Qty){
 OrderNo<-system2(paste0(ExecPath,'Order.exe') ,args=paste(Product,BorS,'0',Qty,'MKT','IOC','1'),stdout = TRUE)
 Match<-system2(paste0(ExecPath,'GetAccount.exe'),args=paste(OrderNo),stdout = TRUE)
 return(Match)
}

#限價委託單
OrderLMT<-function(Product,BorS,Price,Qty){
 OrderNo<-system2(paste0(ExecPath,'Order.exe')  ,args=paste(Product,BorS,Price,Qty,'LMT','ROD','1'),stdout = TRUE)
 return(OrderNo)
}

#單筆帳務查詢
QueryOrder<-function(OrderNo){
 Match<-system2(paste0(ExecPath,'GetAccount.exe'),args=paste(OrderNo),stdout = TRUE)
 return(Match)
}

#總帳務查詢
QueryAllOrder<-function(){
 Match<-system2(paste0(ExecPath,'GetAccount.exe'),args=paste("ALL"),stdout = TRUE)
 return(Match)
}

#未平倉查詢
QueryOnOpen<-function(){
 onopeninfo<-system2(paste0(ExecPath,'OnOpenInterest.exe'),stdout = TRUE)
 return(onopeninfo)
}

#權益數查詢
QueryRight<-function(){
 rightinfo<-system2(paste0(ExecPath,'FutureRights.exe'),stdout = TRUE)
 return(rightinfo)
}

#取消委託單
CancelOrder<-function(OrderNo){
 system2(paste0(ExecPath,'Order.exe')  ,args=paste('Delete',OrderNo),stdout = TRUE)
}

#查詢是否成交
QueryMatch<-function(OrderNo){
 Match<-system2(paste0(ExecPath,'GetAccount.exe'),args=paste(OrderNo),stdout = TRUE)
 if(Match=="Nodata"){
  return(FALSE)
 }else{
  return(TRUE)
 }
}

#限價單到期轉市價單
LMT2MKT <- function(Product,BorS,Price,Qty,Sec){
 OrderNo<-system2(paste0(ExecPath,'Order.exe')  ,args=paste(Product,BorS,Price,Qty,'LMT','ROD','1'),stdout = TRUE)
 to <- Sys.time()
 while(as.numeric(difftime(Sys.time(), to, u = 'secs')) < Sec){
  if(isTRUE(QueryMatch(OrderNo))){
   return(QueryOrder(OrderNo))
  }
 }
 CancelOrder(OrderNo)
 Match<-OrderMKT(Product,BorS,Qty)
 return(Match)
}

#限價單到期轉刪單
LMT2DEL <- function(Product,BorS,Price,Qty,Sec){
 OrderNo<-system2(paste0(ExecPath,'Order.exe')  ,args=paste(Product,BorS,Price,Qty,'LMT','ROD','1'),stdout = TRUE)
 to <- Sys.time()
 while(as.numeric(difftime(Sys.time(), to, u = 'secs')) < Sec){
  if(isTRUE(QueryMatch(OrderNo))){
   return(QueryOrder(OrderNo))
  }
 }
 CancelOrder(OrderNo)
 return(FALSE)
}

File: haohaninfo/即時範例/order.py
# -*- coding: UTF-8 -*-
#載入相關套件
import subprocess

#下單子程式放置位置
ExecPath="./bin/"

#市價單下單
def OrderMKT(Product,BS,Qty):
 OrderNo=subprocess.check_output([ExecPath+"order.exe",Product,BS,"0",Qty,"MKT","IOC","0"]).strip('\r\n')
 while True:
  ReturnInfo=subprocess.check_output([ExecPath+"GetAccount.exe",OrderNo]).strip('\r\n').split(',')
  if len(ReturnInfo)>1:
   return ReturnInfo

#限價單委託
def OrderLMT(Product,BS,Price,Qty):
 OrderNo=subprocess.check_output([ExecPath+"order.exe",Product,BS,Price,Qty,"LMT","ROD","0"]).strip('\r\n')
 return OrderNo

#查詢帳務明細
def QueryOrder(Keyno): 
 ReturnInfo=subprocess.check_output([ExecPath+"GetAccount.exe",Keyno]).strip('\r\n')
 return ReturnInfo.split(',')

#查詢總帳務明細
def QueryAllOrder(): 
 ReturnInfo=subprocess.check_output([ExecPath+"GetAccount.exe","ALL"]).strip('\r\n').split('\r\n')
 ReturnInfo= [ line.split(',') for line in ReturnInfo]
 return ReturnInfo

#查詢未平倉資訊
def QueryOnOpen(): 
 ReturnInfo=subprocess.check_output([ExecPath+"OnOpenInterest.exe"]).strip('\r\n')
 return ReturnInfo.split(',')

#查詢權益數資訊
def QueryRight(): 
 ReturnInfo=subprocess.check_output([ExecPath+"FutureRights.exe"]).strip('\r\n')
 return ReturnInfo.split(',')

#取消委託
def CancelOrder(Keyno):
 ReturnInfo=subprocess.check_output([ExecPath+"order.exe","Delete",Keyno])
 if "cancel send" in ReturnInfo:
  return True
 else:
  return False 

#限價轉刪單
def LMT2DEL(Product,BS,Price,Qty,Sec):
 OrderNo=OrderLMT(Product,BS,Price,Qty)
 StartTime=time.time()
 while time.time()-StartTime<Sec:
  ReturnInfo=QueryOrder(OrderNo)
  if len(ReturnInfo)!=1:
   return ReturnInfo
 CancelOrder(OrderNo)
 return False  

#限價轉市價
def LMT2MKT(Product,BS,Price,Qty,Sec):
 OrderNo=OrderLMT(Product,BS,Price,Qty)
 StartTime=time.time()
 while time.time()-StartTime<Sec:
  ReturnInfo=QueryOrder(OrderNo)
  if len(ReturnInfo)!=1:
   return ReturnInfo
 if CancelOrder(OrderNo):
  ReturnInfo=OrderMKT(Product,BS,Qty)
  return ReturnInfo
